typing/jkind_types.ml

```diff
@@ -578,10 +578,11 @@ type ('type_expr, 'd) t =
   }
 
 module Const = struct
-  type 'type_expr t =
+  type ('type_expr, +'d) t =
     { layout : Layout.Const.t;
       modes_upper_bounds : Modes.t;
```

Comment:
I think it's worthwhile to pass down the `'d` parameter to `Modes.t`, to indicate that this is the part that allowance describes. The current `Modes.t` will probably just ignore this `'d`, but that's still more clearer.

(This code might have been moved around and renamed around "Layout_and_axes", but my point stays.

Reply:
I don't think I agree here. It's true that the allowance describes the modal component, but not in a way that really makes sense at this point in time: it's describing whether or not there are `with`-types, but that's not implemented yet. It does *not* describe left-modes vs right-modes, which is what attaching `'d` to `Modes.t` would look like it means (to me).

(The `'d` also equally applies to the non-mode axes.)

I don't see an easy way to express the valid point you're trying to make here -- that the `'d` is not about the layout! -- but it all becomes much better in the `with`-kinds PR, already posted. So I think it's just easiest to let this one lie, as it's very temporary.

Reply:
> It does not describe left-modes vs right-modes, which is what attaching 'd to Modes.t would look like it means (to me).

Indeed, that should be renamed to `Mode_crossing.t`. Then, it makes sense to talk about left and right, in the following sense: we can compare two mode crossings, and `m0 <= m1` iff `m0` crosses more modes than `m1`.

There are, by my experience with left and right modes, two motivations to distinguish left and right:
- Internally, modes might be join or meet, which inherently forbids them to be on the LHS or RHS of submode, as that would be unsolvable.
- Externally, modes are used as actual or expected, in which case it's healthy to reflect that in the types, by mapping them to left and right modes.

Currently, without the `with` syntax, we don't have the internal motivation. But the external motivation is still valid.

On the other hand, I also understand this might be some work, with benenfits that are not obvious right now. So I'm happy to resolve this conversation if you can just add a CR there.

Reply:
I think you're right to question this, but I don't wish to add the CR. This code changes a bunch in subsequent PRs and the issue you highlight goes away completely once we have `with`-kinds, which is already posted. So I think adding this CR will just cause a bit of work in dealing with conflicts and rebasing but without much benefit.